pkgs/fc/check-ceph/nautilus/fc/check_ceph/check_snapshot_restore.py

```diff
@@ -0,0 +1,251 @@
+                # wording: "could", because we do not know the actual allocation ratio
+                # of the snapshot, the calculation considers the worst case of 100% of
+                # all snapshot blocks being allocated *and* not shared with the main
+                # image (or image being flattened afterwards).
+                f"with {snap['size_bytes'] / GiB :,.2f}GiB might result in a fill "
+                f"ratio of {fill_ratio_after_restore * 100 :.2f}%, "
+                f"exceeding the FULL threshold in cluster root {root}."
+            )
+            exit_code = max_exitcode(exit_code, EXIT_CRITICAL)
+        elif fill_ratio_after_restore >= thresholds.nearfull:
+            print(
+                f"WARN: Restoring the snapshot "
+                f"{pool}/{snap['imgname']}@{snap['snapname']} "
+                f"with {snap['size_bytes'] / GiB :,.2f}GiB might result in a fill "
+                f"ratio of {fill_ratio_after_restore * 100 :.2f}%, "
+                f"exceeding the NEAR_FULL threshold in cluster root {root}."
+            )
+            exit_code = max_exitcode(exit_code, EXIT_WARN)
+        else:
+            print(
+                "OK: Restoring the snapshot "
+                f"{pool}/{snap['imgname']}@{snap['snapname']} "
+                f"with {snap['size_bytes'] / GiB :,.2f}GiB could just increase the "
+                f"fill ratio in cluster root {root} to "
+                f"{fill_ratio_after_restore * 100 :.2f}%."
+            )
+    return exit_code
+
+
+def get_largest_snaps(connection: rados.Rados, pool_roots: dict) -> dict:
+    """Retrieves the largest snapshot per cluster root over all RBD pools specified for
+    that root in pool_roots.
+    The result is a dict of the form {cluster_name: (poolname, snapshotdata)}."""
+    largest_snaps = {}
+    for root, pools in pool_roots.items():
+        # Right now, we only need to deal with a single rbd pool per root, but let's
+        # keep this generic. As fill alerting happens at cluster root level, we need
+        # to find the largest rbd snapshot from all pools of that cluster.
+        root_largest_snap = root_dummy = ("dummypool", {"size_bytes": -1})
+        for pool in pools:
+            with contextlib.closing(connection.open_ioctx(pool)) as poolio:
+                pool_largest_snap = largest_snap_per_pool(poolio, pool)
+                if not pool_largest_snap:
+                    print(f"INFO: Pool {pool} has no (largest) snapshots.")
+                elif (
+                    pool_largest_snap["size_bytes"]
+                    > root_largest_snap[1]["size_bytes"]
+                ):
+                    root_largest_snap = (pool, pool_largest_snap)
+
+        if root_largest_snap is not root_dummy:
+            largest_snaps[root] = root_largest_snap
+    return largest_snaps
+
+
+def largest_snap_per_pool(poolio: rados.Ioctx, poolname: str):
+    largest_snap = dummy = {"size_bytes": -1}  # a pseudo-optional value
```

Comment:
This is also a candidate for splitting between io-oriented code and the actual processing.